### PR TITLE
Add macro to generate register structs, and apply it to nrf52/clock (Fix #1400)

### LINKED ISSUE #1400: [RFC] New macro to generate register structs more safely and easily
# Rationale

One pain point when defining structs of registers for chips is that one needs to manually check that each field ends up at the address defined on the data sheet, and compute the necessary padding between some fields.

This is a quite tedious and error-prone task, which can lead to:
- spending extra time to debug and double-check that all the padding is correct,
- or worse, to incorrect structures making it to the code, if a misaligned field wasn't caught by any test.

A quick `git grep _reserved` shows many padding fields throughout `chips/`:
- Some use arrays of `u8`, some use arrays of `u32`.
- A few use `ReadOnly<u8/u32>` to prevent writing inside the padding. But reserved fields shouldn't even be readable as this could cause a fault too, it should be `NoAccess<u8/u32>` or something like that instead.
- Some have comments explaining the length. But how many bugs have made it among these hundreds of padding fields?

# Proposal

Similarly to the `register_bitfields` macro to manipulate bits within registers, there could be a `register_structs` that:
- Takes as input the address of each field (so it's easier to compare to a data sheet).
- Automatically computes the size of each padding.
- Checks that the addresses are correct, given the size of each field.

## API

I wrote a simple proof-of-concept with `macro_rules`, and the following API should work. If one takes `chips/nrf52/src/adc.rs`, the macro invocation would look like the following:

```rust
register_structs! {
    AdcEventChRegisters [
        (0x000 => limith: ReadWrite<u32, EVENT::Register>)
        (0x004 => limitl: ReadWrite<u32, EVENT::Register>)
        (0x008 => END)
    ],
    AdcChRegisters [
        (0x000 => pselp: ReadWrite<u32, PSEL::Register>)
        (0x004 => pseln: ReadWrite<u32, PSEL::Register>)
        (0x008 => config: ReadWrite<u32, CONFIG::Register>)
        (0x00C => limit: ReadWrite<u32, LIMIT::Register>)
        (0x010 => END)
    ],
    AdcRegisters [
        /// Start the ADC and prepare the result buffer in RAM
        (0x000 => tasks_start: WriteOnly<u32, TASK::Register>)
        /// Take one ADC sample, if scan is enabled all channels are sampled
        (0x004 => tasks_sample: WriteOnly<u32, TASK::Register>)
        /// Stop the ADC and terminate any on-going conversion
        (0x008 => tasks_stop: WriteOnly<u32, TASK::Register>)
        /// Starts offset auto-calibration
        (0x00C => tasks_calibrateoffset: WriteOnly<u32, TASK::Register>)
        (0x010 => _reserved0)
        /// The ADC has started
        (0x100 => events_started: ReadWrite<u32, EVENT::Register>)
        ...
        /// Last result is equal or above `CH[X].LIMIT`
        (0x118 => events_ch: [AdcEventChRegisters; 8])
        (0x158 => _reserved1)
        /// Enable or disable interrupt
        (0x300 => inten: ReadWrite<u32, INTEN::Register>)
        ...
        /// Number of 16 bit samples recorded to RAM
        (0x634 => result_amount: ReadWrite<u32, RESULT_AMOUNT::Register>)
        (0x638 => END)
    ]
}
```

As you can see:
- Each field looks like `(ADDR => field_name: field_type)`. This supports arbitrary complex types (ReadWrite/WriteOnly, arrays, custom types).
- Comments are allowed between fields.
- Each padding looks like `(ADDR => padding_name)`. No need to specify the type nor the length. This saves time computing & debugging the lengths. And a consistent type can be used under the hood.
- The end is marked with `(ADDR => END)`.

This is a proof-of-concept, please let me know your suggestions to improve the syntax (provided that isn't too hard to implement with Rust macros).

## Testing

Besides generating the underlying struct, tests of the form `assert_eq!(core::mem::size_of::<field_type>(), ADDR_NEXT - ADDR, "Field {} has not the expected size", field_name)` can be automatically generated. Once the CI supports running tests on chips (e.g. with https://github.com/tock/tock/pull/1393), this will give feedback to developers adding new registers. Padding lengths won't be wrong anymore, but this can catch errors like forgetting a field.

# Known limitations

## Syntax

- Due to the difficulty of generating identifiers with `macro_rules` (https://stackoverflow.com/questions/27415011/can-a-rust-macro-create-new-identifiers), one still needs to spell out a name for each padding field, instead of using a generic name such as `PADDING`.

- Calling a reserved field `END` would be ambiguous. If that's an issue, the end could be marked as some non-ambiguous symbol like `@END`. Or the padding could be more explicit e.g. `PADDING(_reserved0)`.

## Tests

- Rust doesn't have proper support for `static_assert` (https://internals.rust-lang.org/t/idea-static-assertions/7753). There is a [static_assertions](https://crates.io/crates/static_assertions) crate, but it cannot provide error messages.

I suggest to use a `#[cfg(test)]` gated `#[test]` function, that contains assertions with proper messages (to tell on which field the address is wrong). However, this requires CI running those tests (see https://github.com/tock/tock/pull/1393).

- I've seen some examples where the register type contains a pointer directly (e.g. https://github.com/tock/tock/blob/master/chips/nrf52/src/adc.rs#L58 is `result_ptr: VolatileCell<*const u16>`). As a corollary of the previous point, if the size_of test is done on a desktop that has a different pointer size than the chip (typically 64 vs. 32 bits), then the size test fails on the desktop.

Such pointers don't seem too widespread in the codebase. One solution would be to use a `u32` in the struct and convert that to/from a pointer in the driver code. Another solution would be to define a wrapper type like `ConstPtr32Cell<u16>`, which would contain a `*const u16` on architectures where `#[cfg(target_pointer_width = "32")]` and contain a dummy `u32` with `unimplemented` methods on other architectures. This can generalize to chips that use any pointer size ; within a chip, we know the pointer size anyway.

### PR DESCRIPTION
**Pull request by gendx**
### Pull Request Overview

This pull request adds helper macros to generate register structs, as described in https://github.com/tock/tock/issues/1400. As an example, `chips/nrf52/src/clock.rs` is migrated to use these macros.

The `register_structs` macro does two things:
- Generate the register structs, in particular computing the length of padding fields.
- Create unit tests, checking that each field indeed takes the size between its offset and the next field's offset. These tests require https://github.com/tock/tock/pull/1393 to run (one can run them manually with e.g. `cd chips/nrf52 && CI=true TOCK_KERNEL_VERSION=ci_test cargo test`, although you need to apply the fixes from https://github.com/tock/tock/pull/1393).

Some parts are still not complete, as I'm looking for feedback first.

### Testing Strategy

This pull request was tested by:
- Running `make ci-travis`.
- Running `cargo test` for the nrf52 chip, as described in https://github.com/tock/tock/pull/1393.
- Checking the offsets in the data sheet for the clock at https://infocenter.nordicsemi.com/pdf/nRF52840_OPS_v0.5.pdf (section 17.3).
- Manually double-checking the paddings.
- Flashing the firmware on a nRF52840-DK board and running a simple app (blink).

### TODO or Help Wanted

This pull request still needs:

- [x] Updated `svd2regs.py` script to generate the macro instead of a manual struct.
- [x] Updated documentation for the tock-register-interface library.

I'm waiting for feedback before adding these.

Additionally:

- [x] The generated tests currently require `use super::super::*;` due to putting them in a nested module. Wildcard imports are currently prohibited by Tock style guide. I added an exception for this case.

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make formatall`.

### CODE REVIEW

**Comment by ppannuto** [comment]
This is awesome, +1 here as well once there's documentation added to the register interface.

One maybe-nitpick on the syntax: is it weird to not have commas after each entry? This looks like a list-like thing, my intuition expects commas. While I find it perfectly readable with just newlines as-is, I think commas would be more consistent with rust syntax as a whole?

**Comment by gendx** [comment]
The extra commas make sense. I also updated the syntax to mark the end with `@END`, which stands out more and doesn't conflict with an identifier.

Added the documentation for the register interface.

So the remaining points are `svd2regs.py` and the fact that unit tests currently require a wildcard import (although they won't run on CI until https://github.com/tock/tock/pull/1393 is merged).

**Comment by ia0** [comment]
I'm new to embedded programming, but I wonder why use a `struct` for registers in Rust. The API we want is given a register name, get access to that register. This can indeed be encoded with the field offsets of a repr(C) struct, but it can also be encoded using constant functions (producing the same assembly). Those fake structs don't exist, we never allocate them. They are just a way to encode an offset and provide a way to add it to a base address. This can be done with a constant function too.

To illustrate consider this macro to generate registers:
```rust
macro_rules! define_registers {
    (struct $Name: ident { $($name: ident: $type: ty = $offset: expr),*$(,)? }) => {
        #[derive(Clone, Copy)]
        struct $Name { base_address: usize }
        impl $Name {
            $(const fn $name(&self) -> StaticRef<$type> {
                unsafe { StaticRef::new((self.base_address + $offset) as *const $type) }
            })*
        }
        #[test]
        #[allow(non_snake_case)]
        fn $Name() {
            struct Field { name: &'static str, offset: usize, size: usize }
            let mut fields = [$(Field {
                name: stringify!($name),
                offset: $offset,
                size: core::mem::size_of::<$type>()
            }),*];
            fields.sort_by_key(|field| field.offset);
            for window in fields.windows(2) {
                let (prev, next) = (&window[0], &window[1]);
                assert!(prev.offset + prev.size <= next.offset,
                        "{} and {} overlap in {}",
                        prev.name, next.name, stringify!($Name));
            }
        }
    };
}
```

When given the following input:
```rust
define_registers! {
    struct ClockRegisters {
        tasks_hfclkstart: WriteOnly<u32, Control::Register> = 0x000,
        tasks_hfclkstop: WriteOnly<u32, Control::Register> = 0x004,
        tasks_lfclkstart: ReadWrite<u32, Control::Register> = 0x008,
        ...
        lfclksrccopy: ReadOnly<u32, LfClkSrcCopy::Register> = 0x41C,
        lfclksrc: ReadWrite<u32, LfClkSrc::Register> = 0x518,
        ctiv: ReadWrite<u32, Ctiv::Register> = 0x538,
        traceconfig: ReadWrite<u32, TraceConfig::Register> = 0x55C,
    }
}
```

It produces the following output:
```rust
struct ClockRegisters { base_address: usize }
impl ClockRegisters {
    const fn tasks_hfclkstart(&self) -> StaticRef<WriteOnly<u32, Control::Register>> {
        unsafe { StaticRef::new((self.base_address + 0x000) as *const WriteOnly<...>) }
    }
    const fn tasks_hfclkstop(&self) -> ...
    ...
}
```

And a test named `ClockRegisters` that checks if any 2 fields (fields don't need to be ordered by offset in the macro input) are overlapping and fails with the following error message: `<prev_field> and <next_field> overlap in ClockRegisters`.

The main difference with what is currently done in Tock (and what this PR continues to do) is that accessing a register is done through a constant function call rather than a field access:
```rust
// Instead of
let regs = &*self.registers;
regs.intenset.write(...)
// It is
let regs = self.registers;
regs.intenset().write(...)
```

Also, initializing the registers is different and has a different type:
```rust
// Instead of
const CLOCK_BASE: StaticRef<ClockRegisters> =
    unsafe { StaticRef::new(0x40000000 as *const ClockRegisters) };
// It is
const CLOCK_BASE: ClockRegisters = ClockRegisters {
    base_address: 0x40000000,
};
```

Maybe this approach doesn't make sense, but it seems to me more adapted to Rust. The assembly should be exactly the same. So it's mostly a difference of APIs:
- As input, we provide for each register: name, type, offset. Without the need to mention and name padding (which is the only significant difference). Overlapping fields are still detected.
- As output, we get access to the register by calling a constant function instead of using a field.

Is there a reason I'm missing why structs are used in Tock to represent registers?

**Comment by bradjc** [comment]
@ia0 I think the answer to your question from Tock's point of view is based on this generated piece of code:

```rust
const fn tasks_hfclkstart(&self) -> StaticRef<WriteOnly<u32, Control::Register>> {
    unsafe { StaticRef::new((self.base_address + 0x000) as *const WriteOnly<...>) }
}
```

Since a macro is creating that, chip drivers are using `unsafe` without the string "unsafe" appearing anywhere in the code. This type of hiding is pretty opposite of Tock's goals. With the struct approach we can be explicit that the unsafe operation is casting the MMIO pointer to a specific struct. Assuming that a developer has check that that address is correct, and the struct is correct, then that is memory safe. It also is grepable and easier to audit.

**Comment by gendx** [comment]
@bradjc I briefly mentioned the reason in the issue https://github.com/tock/tock/issues/1400:

> Due to the difficulty of generating identifiers with macro_rules (https://stackoverflow.com/questions/27415011/can-a-rust-macro-create-new-identifiers), one still needs to spell out a name for each padding field, instead of using a generic name such as PADDING.

According to the StackOverflow thread, generating unique identifiers isn't an easy task with Rust macros, due to a lack of built-in support in the compiler. There seem to be a `mashup` crate that could allow that, but I'm not sure it's worth adding this extra dependency, nor how intuitive the syntax will end up.

If you thing it's worth it, the syntax could be `0x060 => @RESERVED(_reserved0)` instead of `0x060 => _reserved0`. I personally find that the extra verbosity is not worth it, but it wouldn't be hard to update the macros to support that.

**Comment by gendx** [comment]
Ok, I'll figure out how to disable the check to allow `super::super::*` in this case.

**Comment by ia0** [comment]
@bradjc I see, it makes a lot of sense. Since the question is only about making the use of unsafe apparent (the macro definition has to be reviewed in any circumstances due to the [scope of unsafe][1]), it is easy to force the user to pay attention by forcing them to mention the `unsafe` keyword where something has to be checked.

First, the input of the macro must be checked (register names, offsets, and types):
```rust
define_registers! {
    unsafe struct ClockRegisters {
        tasks_hfclkstart: WriteOnly<u32, Control::Register> = 0x000,
        tasks_hfclkstop: WriteOnly<u32, Control::Register> = 0x004,
        tasks_lfclkstart: ReadWrite<u32, Control::Register> = 0x008,
```

Second, the base address has to be checked:
```rust
const CLOCK_BASE: ClockRegisters = ClockRegisters {
    base_address: unsafe { BaseAddress::new(0x40000000) },
};
```

Here is how the implementation looks like (omitting the glue):
```rust
// libraries/tock-register-interface/src/lib.rs
#[derive(Clone, Copy)]
pub struct BaseAddress {
    base_address: usize,
}
impl BaseAddress {
    pub const unsafe fn new(base_address: usize) -> BaseAddress {
        BaseAddress { base_address }
    }
    pub const fn offset(&self, offset: usize) -> usize {
        self.base_address + offset
    }
}

// libraries/tock-register-interface/src/macros.rs
#[macro_export]
macro_rules! define_registers {
    (unsafe struct $Name: ident { $($name: ident: $type: ty = $offset: expr),*$(,)? }) => {
        #[derive(Clone, Copy)]
        struct $Name { base_address: $crate::BaseAddress }
        impl $Name {
            $(const fn $name(&self) -> StaticRef<$type> {
                unsafe { StaticRef::new(self.base_address.offset($offset) as *const $type) }
            })*
        }
        // Omitting the test generation which is identical to before.
    };
}
```

[1]: https://www.ralfj.de/blog/2016/01/09/the-scope-of-unsafe.html

**Comment by rajivr** [comment]
Firstly this is really awesome! :-) 

I just tested this. I am running into two issues.

Following form compiles correctly.

```
register_structs!{
    UsbRegisters [
        (0x000 => usb_inten: ReadWrite<u32, USB_INTEN::Register>),
	...
	(0xXXX => @END),
    ]
}
```

However, when generating code from `svd2regs.py`, two things happen.

1. When comments are included (e.g., `/// USB Register Map`), it fails to compile. `svd2regs.py` generates comments by default.

```
register_structs!{
    /// USB Register Map
    UsbRegisters [
        (0x000 => usb_inten: ReadWrite<u32, USB_INTEN::Register>),
	...
	(0xXXX => @END),
    ]
}
````

2. It seems like `[` and `{` is being interchanged by `svd2regs.py`. Following form is generated that fails.

```
register_structs![
    /// USB Register Map
    UsbRegisters {
        (0x000 => usb_inten: ReadWrite<u32, USB_INTEN::Register>),
	...
	(0xXXX => @END),
    }
]
```

**Comment by ia0** [comment]
Another point against using structs is that it relies on `repr(C)`. This is probably not an issue in practice, but it's one more dependency that can go wrong. This could be fixed in the current PR by checking, in the generated test, that the field offsets in the output struct are actually equal to their offset in the macro input.

**Comment by gendx** [comment]
@rajivr Thanks for the feedback! I updated the syntax to use curly braces everywhere in the macro, this should avoid confusion with square brackets. The macro should also accept comments before structs.

@ia0 That would be a great idea, but it seems that the `offset_of` macro in Rust's standard library is still under discussion: https://internals.rust-lang.org/t/pre-rfc-add-a-new-offset-of-macro-to-core-mem/9273
I'd wait for it to settle down before integrating it here.
In the meantime, there is a sizeof check for the whole struct, besides sizeof and alignof checks for individual fields, so if `repr(C)` somehow yields something unexpected it's likely to be caught by these tests.

If someone is more expert than me on Bash and can figure out a way to easily white-list `src/macro.rs` for wildcard imports **only** in `libraries/tock-register-interface/` that would be awesome!

**Comment by gendx** [comment]
@bradjc I updated a sentence that I didn't find very clear in the readme.

**Comment by ia0** [comment]
@gendx Right, the existing checks make it already very unprobable that the offsets would be wrong. Adding such checks is unnecessary.

**Comment by rajivr** [comment]
> @rajivr Thanks for the feedback! I updated the syntax to use curly braces everywhere in the macro, this should avoid confusion with square brackets. The macro should also accept comments before structs.

Thanks for the update. I am out all day today. I'll try to test it later this evening or tomorrow.

**Comment by bradjc** [comment]
> @bradjc I updated a sentence that I didn't find very clear in the readme.

Is it not true that gaps become arrays? Does this make an array of `cr` that is 8 entries long?

```rust
register_structs! {
    Registers {
        (0x000 => cr: ReadWrite<u8, Control::Register>),
        (0x008 => s: ReadOnly<u8, Status::Register>),
    }
}
```

In either case the readme should specify what happens. I thought based on how reserved work that it would create an array.

**Comment by ppannuto** [comment]
I don't think gaps can automatically become arrays, because the array would
need a name ("_reservered2" or whatever), and there isn't a way to create
names in macros right now

**Comment by gendx** [comment]
No, your example would try to create:

```rust
#[repr(C)]
struct Registers {
    cr: ReadWrite<u8, Control::Register>,
    s: ReadOnly<u8, Status::Register>,
}
```

and then unit tests which would essentially do:

```rust
let mut sum = 0;
assert_eq(sum, 0);
sum += size_of::<u8>();
assert_eq(sum, 8);
sum += size_of::<u8>();
...
```

So the assert would break, but that would be detected by CI only after https://github.com/tock/tock/pull/1393 is merged.

It is however true that gaps becomes arrays of `u8` (ideally it should be something like `NoAccess<u8>` though), but gaps require an explicit identifier such as `_reserved0`.

**Comment by rajivr** [comment]
I tested the latest patch. It look good! :+1:

**Comment by bradjc** [comment]
So doing this:

https://github.com/tock/tock/blob/69c3a5d5583468a7775c5f193e7d05b7b1769755/chips/sam4l/src/acifc.rs#L84

would be:

```rust
register_structs! {
    Registers {
        (0x000 => [cr: ReadWrite<u8, Control::Register>; 8]),
        (0x008 => s: ReadOnly<u8, Status::Register>),
    }
}
```

**Comment by gendx** [comment]
You can use any type as a "register", including an array type - in this case the type is `[ReadWrite<u32, WindowConfiguration::Register>; 4]`.

So here you'd need something like:

```rust
register_structs! {
AcifcRegisters {
    (0x000 => ctrl: ReadWrite<u32, Control::Register>),
    ...
    (0x038 => _reserved2),
    (0x080 => confw: [ReadWrite<u32, WindowConfiguration::Register>; 4]),
    (0x090 => _reserved3),
    (0x0D0 => conf: [ReadWrite<u32, ACConfiguration::Register>; 8]),
    (0xXXX => @END),
}
}
```

It's actually similar to the regular syntax for structs, apart from the explicit addresses, and the implicit type for gaps.

I can add this case to the examples if it makes it clearer.

File: tools/run_cargo_fmt.sh [hunk 1]
```diff
@@ -64,7 +64,7 @@ done
 # Find folders with Cargo.toml files in them and check them (avoids matching this script!)
 for f in $(find . | grep Cargo.toml); do
 	pushd $(dirname $f) > /dev/null
-	if $(git grep -q 'use .*\*;'); then
+	if $(git grep -q 'use .*\*;' -- ':!src/macros.rs'); then
```
Comment (bradjc) [inline tools/run_cargo_fmt.sh]: Why does macros.rs need to be excluded?

I prefer to trust in rustfmt, but there are ways to be very specific about what fustfmt should ignore.

**Comment by bradjc** [comment]
Ok gotcha, and thanks for adding that to the readme.

One last question: so only entries with a single name (`(0x090 => _reserved3)`) automatically become arrays, or only things that start with `_reserved`? Put another way, if someone specifies the space as `(0x090 => padding)` will that work?

File: tools/run_cargo_fmt.sh (hunk 1 above)
Comment (gendx) [reply tools/run_cargo_fmt.sh]: The confusion part here is that this `run_rust_fmt.sh` script also handles looking for wildcard import patterns (the part that I changed here), which has nothing to do with formatting. I'd suggest to split the script in 2 files to make this clearer.

**Comment by gendx** [comment]
The macro doesn't actually check the identifier, you can use whatever you want, including "padding". I would say that using `_reservedN` would be more of a style convention within Tock.

Note that since such fields are never used (or at least they shouldn't!), not pre-pending them with an underscore will lead the compiler to complain about dead code. And you cannot reuse twice the same identifier in a struct, so if you have multiple gaps, you'd need to name them something like `padding0`, `padding1`, etc. The simplest is probably to use `_reservedN` (which is also output by `svd2regs.py`).

File: tools/run_cargo_fmt.sh (hunk 1 above)
Comment (gendx) [reply tools/run_cargo_fmt.sh]: Moved that part to a standalone script.

**Comment by bradjc** [comment]
But it is based on just having an identifier? I'm still trying to understand why in one case an array is automatically created and in my earlier case an array is not automatically created and instead a gap is introduced.

As for the naming, I just want to make sure I understand so I can help users if they have questions in the future.

**Comment by gendx** [comment]
The syntax is defined as follows in the macro:

- `($offset_start:expr => $field:ident: $ty:ty)` yields `$field: $ty` in the struct.
- `($offset_start:expr => $padding:ident)` yields `$padding: [u8; $offset_end - $offset_start]` in the struct (where `$offset_end:expr` is obtained from the next field).

So it is based on having a type (separated by a colon) or not. In all cases an identifier is required, as there's no easy way to generate one in a Rust macro (as discussed in the issue).

If you find that confusing, I could make the syntax more explicit, e.g. `0x009 => @PADDING(_reserved0)`.

**Comment by bradjc** [comment]
Ok I think I understand now. Please double check my updates to the readme, but I think they are consistent with the code and should give people enough detail to effectively use the register_structs macro.

**Comment by andre-richter** [comment]
That sounds incredibly useful. Thanks!

One thing I noticed is that the README is kind of a roller-coaster where you first think that the no.1 selling point of this addition, naemly automatic gap calculation, isn't there. Readers will first stumble upon this sentence:

```
Any gap when defining the registers will **not** be preserved when the register struct
is created by the macro. The macro also generates a unit test which will fail if a gap
is introduced.
```

I read that and was like, bummer, I still have to manually fill. Only the next few sentences explain it, but still its not 100% clear.

I would put it something like:
"Gaps when defining the registers must be explicitly annotated (by convention using `_reservedN`), the macro will then automatically take care of calculating the gap size and inserting a respective filler struct.

**Comment by gendx** [comment]
@andre-richter I think the original formulation in the documentation had a more direct explanation. Thanks for your proposal, this sounds clear!

**Comment by gendx** [comment]
I added an example of unintentional gap. I hope that this is clear enough (and that the discussion on https://github.com/tock/tock/pull/1393 will settle down).

**Comment by alevy** [comment]
So just waiting on @bradjc to approve I think

**Comment by ia0** [comment]
FYI: I created #1439 to continue discussion on avoiding encoding the register name to offset mapping in the field offsets of a `repr(C)` struct. This should not block merging this PR which is already improving the situation.

**Comment by gendx** [comment]
Looks like this is still not merged by bors.

**Pull request merged**